# __Group Project 2 - Team Wilbur Atwater__
<br>


### __Team Members: Lara Gursoy (Project Manager), Himalia Joshi, Qianhao Zhong, Bolor Erdene-Ochir, Moria Zhut__
<br>


## **1. Project Topic:**

> __Examining the minimum subsistence price for 3 different kinds of diets (keto, mediterranean, vegan) using one list of products, adjusting the function for each diet to put different constrictions on nutrition in accordance with each diet.__
<br>


## **2. Project Goal:**
> * __Find which diet is the most expensive and least expensive.__
> * __Find which diet is the least "nutritious" in comparison to federal nutrition recommendations and what its price is.__
> * __Compare our findings of each diet with preconceived notions/widespread opinions about each diet (eg: vegan is expensive/unhealthy).__
<br>


## **3. [A]Population of Interest:**
> __As Berkeley students ourselves, we wanted to see the minimum cost diets of possible college students that may be on these various diets. The main grocery stores around campus are Trader Joe’s, Safeway, and some small corner stores. Students most likely shop mostly at Trader Joe’s and Safeway as they have the most options and the more affordable prices. We chose Safeway as it has more affordable options than Trader Joe’s. We are looking at the nutritional requirements for females in the age range 19-30 (around 56% of UC Berkeley’s student population is female).__

<br>


## **4. Project Deliverables:**

## Packages Import:

In [1]:
import pandas as pd
import numpy as np
import fooddatacentral as fdc
import warnings
from  scipy.optimize import linprog as lp
import cufflinks as cf

### [A] Dietary Reference Intakes:

In [42]:
# This part is to create a panda.Series of dietary requirements.
gender = input("Gender(C or F or M): ").upper()
age_range = input("Age Range: ")  # Format of age_range: "M 31-50", "F 51+", "C 1-3"
max_or_min = input("Max or Min: ").lower()
min_requirement = pd.read_csv("Dietary_Requirements_diet_minimums.csv")
min_requirement = min_requirement.drop(min_requirement.columns[1], axis = 1)
max_requirement = pd.read_csv("Dietary_Requirements_diet_maximums.csv")
max_requirement = max_requirement.drop(max_requirement.columns[1], axis = 1)

def min_require(age_range, gender):  # return a pandas.Series
    query_char = str(gender) + " " + str(age_range)
    min_series = pd.Series(min_requirement[query_char].tolist(), index = min_requirement["Nutrition"].tolist())
    return min_series

def max_require(age_range, gender):  # return a pandas.Series
    query_char = str(gender) + " " + str(age_range)
    max_series = pd.Series(max_requirement[query_char].tolist(), index = max_requirement["Nutrition"].tolist())
    return max_series

if max_or_min == "min":
    print("\n")
    print(min_require(age_range, gender))
    print("\n")
    print(type(min_require(age_range, gender)))
elif max_or_min == "max":
    print("\n")
    print(max_require(age_range, gender))
    print("\n")
    print(type(max_require(age_range, gender)))

Gender(C or F or M): m
Age Range: 31-50
Max or Min: min


Energy                            2200.0
Protein                             56.0
Fiber, total dietary                30.8
Folate, DFE                        400.0
Calcium, Ca                       1000.0
Carbohydrate, by difference        130.0
Iron, Fe                             8.0
Magnesium, Mg                      420.0
Niacin                              16.0
Phosphorus, P                      700.0
Potassium, K                      4700.0
Riboflavin                           1.3
Thiamin                              1.2
Vitamin A, RAE                     900.0
Vitamin B-12                         2.4
Vitamin B-6                          1.3
Vitamin C, total ascorbic acid      90.0
Vitamin E (alpha-tocopherol)        15.0
Vitamin K (phylloquinone)          120.0
Zinc, Zn                            11.0
dtype: float64


<class 'pandas.core.series.Series'>


### [A] Data on Prices for Difference foods:

In [43]:
# This part provides the pandas.DataFrame version of the csv file.
food_price = pd.read_csv("Food_Price_Data.csv")
food_price = food_price.drop(food_price.columns[5:], axis = 1)
food_price

fdcId                    Name  Quantity      Unit  Price ($)
0   2120006  chicken breast tenders      1.00        lb       7.99
1   2390897  Atlantic salmon fillet      1.00        lb      10.99
2   2094567    Cage free large eggs      1.00     dozen       5.99
3   2344719                 Avocado      5.00        oz       2.00
4   2577037  Extra virgin olive oil     17.00  fluid oz       7.99
5   2261421       Oat - Whole Grain     42.00        oz       6.99
6   2345841                   Honey     12.00        oz       5.99
7   2067508      Greek Yogurt Plain     32.00        oz       6.99
8   2608592             Raw Almonds     16.00        oz       7.99
9   2012128                  Banana      1.00        lb       1.99
10  2099349                 Spinach     16.00        oz       5.99
11  1858805                 Quinoa      16.00        oz       3.79
12  2429179         Unsalted Butter     16.00        oz       5.99
13  2670943                 Parsley      0.35        oz       4.49
14  1859624             Ripe olives      2.25        oz       1.99
15  2345210        Sweet Potato/Yam      1.00        lb       1.99
16  2345293           Green Cabbage      1.00        lb       1.49
17  2262072           Peanut Butter     16.30        oz       3.99
18  2345173                 Carrots      1.00        lb       1.49
19  2343189       Whole Wheat Bread     22.00        oz       2.99
20   168877         Long Grain rice     32.00        oz       2.49
21  2570691         Spaghetti Pasta     16.00        oz       2.50
22  2514744             Ground beef      3.00        lb      14.97
23  2101796         Shredded cheese     32.00        oz       9.99
24  2461103           Cheese Slices      8.00        oz       3.99
25  2340762              Whole Milk    128.00  fluid oz       5.99
26   169255    Portabello Mushrooms      1.00        oz       0.50
27  1922954                Tomatoes     10.00        oz       3.69
28  2345304                Cucumber     32.00        oz       5.99
29  2345292        Brussels Sprouts      1.00        lb       3.99
30   168246       Pork Loin Sirloin      7.00        lb      13.93
31  2344777            Strawberries      1.00        lb       5.99
32  2341566         Kidney(chicken)      1.25       lbs       4.99

In [44]:
# This part is used to convert the units, such as "oz" and "lb", into standard units such as "hectogram".
# If it doesn't work, try to run the upper code block first!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

# Unit Conversion/Normalization
food_price["FDC Quantity"] = food_price.apply(lambda x: fdc.units(x["Quantity"], x["Unit"]), axis = 1)
food_price["FDC Price"] = food_price["Price ($)"] / food_price["FDC Quantity"]
warnings.filterwarnings("ignore", message="The unit of the quantity is stripped when downcasting to ndarray")
food_price = food_price.drop(food_price.columns[2:6], axis = 1)  
food_price

fdcId                    Name                        FDC Price
0   2120006  chicken breast tenders   1.7614934748571718 / hectogram
1   2390897  Atlantic salmon fillet   2.4228802614118043 / hectogram
2   2094567    Cage free large eggs                 nan / milliliter
3   2344719                 Avocado   1.4109584779832163 / hectogram
4   2577037  Extra virgin olive oil                 nan / milliliter
5   2261421       Oat - Whole Grain   0.5870595095894454 / hectogram
6   2345841                   Honey    1.760758600649889 / hectogram
7   2067508      Greek Yogurt Plain    0.770515606336147 / hectogram
8   2608592             Raw Almonds   1.7614934748571718 / hectogram
9   2012128                  Banana   0.4387199017479063 / hectogram
10  2099349                 Spinach   1.3205689504874165 / hectogram
11  1858805                 Quinoa    0.8355519736806859 / hectogram
12  2429179         Unsalted Butter   1.3205689504874165 / hectogram
13  2670943                 Parsley     45.2514540438903 / hectogram
14  1859624             Ripe olives    3.119785967985112 / hectogram
15  2345210        Sweet Potato/Yam   0.4387199017479063 / hectogram
16  2345293           Green Cabbage  0.32848877065546755 / hectogram
17  2262072           Peanut Butter   0.8634546514038395 / hectogram
18  2345173                 Carrots  0.32848877065546755 / hectogram
19  2343189       Whole Wheat Bread   0.4794052101329338 / hectogram
20   168877         Long Grain rice   0.2744755164201726 / hectogram
21  2570691         Spaghetti Pasta   0.5511556554621939 / hectogram
22  2514744             Ground beef   1.1001066883025392 / hectogram
23  2101796         Shredded cheese   1.1012089996134633 / hectogram
24  2461103           Cheese Slices    1.759288852235323 / hectogram
25  2340762              Whole Milk                 nan / milliliter
26   169255    Portabello Mushrooms   1.7636980974790204 / hectogram
27  1922954                Tomatoes    1.301609195939517 / hectogram
28  2345304                Cucumber   0.6602844752437083 / hectogram
29  2345292        Brussels Sprouts   0.8796444261176615 / hectogram
30   168246       Pork Loin Sirloin   0.4387199017479063 / hectogram
31  2344777            Strawberries   1.3205689504874165 / hectogram
32  2341566         Kidney(chicken)   0.8800853506420312 / hectogram

### [A] Nutritional Content of Different Foods:

In [45]:
# This part is used to query the fdcId for specific food.
# Get apikey from https://fdc.nal.usda.gov/
food = input("Food Name: ")  # "crunchy peanut butter"
apikey = "mF4KpJl1hVML9jWQcT8iJZYIRXjD67LVCFzrsqN4"
food_key_df = fdc.search(apikey, food)
food_key_df = food_key_df.drop(food_key_df.columns[3:], axis = 1)
food_key_df

Food Name: milk


fdcId     description        dataType
0   1909132            MILK         Branded
1   2454032            MILK         Branded
2   2078244            MILK         Branded
3   2083027            MILK         Branded
4   2062196            MILK         Branded
5   2503416            MILK         Branded
6   2475001            MILK         Branded
7   2501185            MILK         Branded
8   2476573            MILK         Branded
9   2092649            MILK         Branded
10  1909141            MILK         Branded
11  2098011            MILK         Branded
12  2630156            MILK         Branded
13  2658737            MILK         Branded
14  2471393            MILK         Branded
15  1954124            MILK         Branded
16  2062449            MILK         Branded
17  1881028            MILK         Branded
18  2074481            MILK         Branded
19  2085761            MILK         Branded
20  2078442            MILK         Branded
21  2657009            MILK         Branded
22  1909137            MILK         Branded
23  2062204            MILK         Branded
24  2083378            MILK         Branded
25  2658501            MILK         Branded
26  1960596            MILK         Branded
27  2092666            MILK         Branded
28  2481948            MILK         Branded
29  2098490            MILK         Branded
30  1909134            MILK         Branded
31  1909135            MILK         Branded
32  2430701            MILK         Branded
33  2464034            MILK         Branded
34  2079089            MILK         Branded
35  2673630            MILK         Branded
36  1960637            MILK         Branded
37  2089787            MILK         Branded
38  2624794            MILK         Branded
39  2338716            MILK         Branded
40  2118869            MILK         Branded
41  2340762     Milk, whole  Survey (FNDDS)
42  2340794    Coconut milk  Survey (FNDDS)
43  2343642  Crackers, milk  Survey (FNDDS)
44   174980  Crackers, milk       SR Legacy
45  2340760     Milk, human  Survey (FNDDS)
46  2340900    Milk, malted  Survey (FNDDS)
47  2340761       Milk, NFS  Survey (FNDDS)
48  2340793       Rice milk  Survey (FNDDS)
49  2340784        Soy milk  Survey (FNDDS)

In [46]:
# This part provides the nutrient contents and their quantities of the provided food fdcId.
FDCID = str(input("The fdcId you picked from above table: "))
nutrition_df = fdc.nutrients(apikey, FDCID)
nutrition_df

The fdcId you picked from above table: 2340784


Quantity Units
Protein                                3.410     g
Total lipid (fat)                      2.050     g
Carbohydrate, by difference            5.220     g
Energy                                53.000  kcal
Alcohol, ethyl                         0.000     g
...                                      ...   ...
PUFA 20:5 n-3 (EPA)                    0.000     g
MUFA 22:1                              0.000     g
PUFA 22:5 n-3 (DPA)                    0.000     g
Fatty acids, total monounsaturated     0.399     g
Fatty acids, total polyunsaturated     1.100     g

[65 rows x 2 columns]

### [A] Matrix $A$ Construction (Nutrients-Food):

In [47]:
# This part is a function that outputs the matrix A
def matrix_A_func(food_price):
    
    # Normalized units with corresponding prices
    p = food_price.drop(food_price.columns[0], axis = 1)
    p = p[~p['FDC Price'].astype(str).str.contains('nan', case=False, na=False)]
    # astype(str) converts 'FDC Price' column to strings and applies the str.contains method to filter out rows containing 'nan'. 

    p = p.set_index('Name')
    ###############################################################################################################################
    min_requirement = pd.read_csv("Dietary_Requirements_diet_minimums.csv")
    min_requirement = min_requirement.drop(min_requirement.columns[1], axis = 1)
    min_requirement = min_requirement.set_index('Nutrition')
    ###############################################################################################################################
    max_requirement = pd.read_csv("Dietary_Requirements_diet_maximums.csv")
    max_requirement = max_requirement.drop(max_requirement.columns[1], axis = 1)
    max_requirement = max_requirement.set_index('Nutrition')
    ###############################################################################################################################

    matrix_A = {}

    for food in food_price.Name.tolist():
        try:
            FDCID = food_price.loc[food_price.Name == food, :].fdcId.iloc[0]
            matrix_A[food] = fdc.nutrients(apikey, FDCID).Quantity   
        except AttributeError:
            warnings.warn(f"Couldn't find FDC Code {FDCID} for food {food}.")

    matrix_A = pd.DataFrame(matrix_A, dtype = float)
    matrix_A = matrix_A[p.index].fillna(0)

    # Drop the rows that we don't have constraints for.
    matrix_A_min = matrix_A.loc[min_requirement.index]
    matrix_A_max = matrix_A.loc[max_requirement.index]

    # Concatentate matrix_A_min and matrix_A_max
    matrix_A = pd.concat(([matrix_A_min, -matrix_A_max]))
    return matrix_A


matrix_A = matrix_A_func(food_price)
matrix_A

chicken breast tenders  \
Nutrition                                                
Energy                                           89.00   
Protein                                          18.75   
Fiber, total dietary                              0.00   
Folate, DFE                                       0.00   
Calcium, Ca                                       0.00   
Carbohydrate, by difference                       0.00   
Iron, Fe                                          0.32   
Magnesium, Mg                                     0.00   
Niacin                                            0.00   
Phosphorus, P                                     0.00   
Potassium, K                                      0.00   
Riboflavin                                        0.00   
Thiamin                                           0.00   
Vitamin A, RAE                                    0.00   
Vitamin B-12                                      0.00   
Vitamin B-6                                       0.00   
Vitamin C, total ascorbic acid                    0.00   
Vitamin E (alpha-tocopherol)                      0.00   
Vitamin K (phylloquinone)                         0.00   
Zinc, Zn                                          0.00   
Sodium, Na                                     -250.00   
Energy                                          -89.00   

                                Atlantic salmon fillet  Avocado  \
Nutrition                                                         
Energy                                          212.00  160.000   
Protein                                          20.35    2.000   
Fiber, total dietary                              0.00    6.700   
Folate, DFE                                       0.00   81.000   
Calcium, Ca                                       0.00   12.000   
Carbohydrate, by difference                       0.00    8.530   
Iron, Fe                                          0.32    0.550   
Magnesium, Mg                                     0.00   29.000   
Niacin                                            0.00    1.740   
Phosphorus, P                                     0.00   52.000   
Potassium, K                                    248.00  485.000   
Riboflavin                                        0.00    0.130   
Thiamin                                           0.00    0.067   
Vitamin A, RAE                                    0.00    7.000   
Vitamin B-12                                      0.00    0.000   
Vitamin B-6                                       0.00    0.257   
Vitamin C, total ascorbic acid                    0.00   10.000   
Vitamin E (alpha-tocopherol)                      0.00    2.070   
Vitamin K (phylloquinone)                         0.00   21.000   
Zinc, Zn                                          0.00    0.640   
Sodium, Na                                      -58.00   -7.000   
Energy                                         -212.00 -160.000   

                                Oat - Whole Grain    Honey  \
Nutrition                                                    
Energy                                    0.00000  304.000   
Protein                                  13.16875    0.300   
Fiber, total dietary                     10.50000    0.200   
Folate, DFE                               0.00000    2.000   
Calcium, Ca                              42.81000    6.000   
Carbohydrate, by difference              69.91725   82.400   
Iron, Fe                                  3.99700    0.420   
Magnesium, Mg                           125.30000    2.000   
Niacin                                    1.94500    0.121   
Phosphorus, P                           371.80000    4.000   
Potassium, K                            373.10000   52.000   
Riboflavin                                0.16130    0.038   
Thiamin                                   0.39000    0.000   
Vitamin A, RAE                            0.00000    0.000   
Vitamin B-12                          

### [A] Matrix $b$ Construction (Nutrients-Age&Gender)

In [48]:
# Minimum Requirement DataFrame
min_requirement = pd.read_csv("Dietary_Requirements_diet_minimums.csv")
min_requirement = min_requirement.drop(min_requirement.columns[1], axis = 1)
min_requirement = min_requirement.set_index("Nutrition")
min_requirement

C 1-3   F 4-8   M 4-8  F 9-13  M 9-13  \
Nutrition                                                                
Energy                          1000.0  1200.0  1400.0  1600.0  1800.0   
Protein                           13.0    19.0    19.0    34.0    34.0   
Fiber, total dietary              14.0    16.8    19.6    22.4    25.2   
Folate, DFE                      150.0   200.0   200.0   300.0   300.0   
Calcium, Ca                      700.0  1000.0  1000.0  1300.0  1300.0   
Carbohydrate, by difference      130.0   130.0   130.0   130.0   130.0   
Iron, Fe                           7.0    10.0    10.0     8.0     8.0   
Magnesium, Mg                     80.0   130.0   130.0   240.0   240.0   
Niacin                             6.0     8.0     8.0    12.0    12.0   
Phosphorus, P                    460.0   500.0   500.0  1250.0  1250.0   
Potassium, K                    3000.0  3800.0  3800.0  4500.0  4500.0   
Riboflavin                         0.5     0.6     0.6     0.9     0.9   
Thiamin                            0.5     0.6     0.6     0.9     0.9   
Vitamin A, RAE                   300.0   400.0   400.0   600.0   600.0   
Vitamin B-12                       0.9     1.2     1.2     1.8     1.8   
Vitamin B-6                        0.5     0.6     0.6     1.0     1.0   
Vitamin C, total ascorbic acid    15.0    25.0    25.0    45.0    45.0   
Vitamin E (alpha-tocopherol)       6.0     7.0     7.0    11.0    11.0   
Vitamin K (phylloquinone)         30.0    55.0    55.0    60.0    60.0   
Zinc, Zn                           3.0     5.0     5.0     8.0     8.0   

                                F 14-18  M 14-18  F 19-30  M 19-30  F 31-50  \
Nutrition                                                                     
Energy                           1800.0   2200.0   2000.0   2400.0   1800.0   
Protein                            46.0     52.0     46.0     56.0     46.0   
Fiber, total dietary               25.2     30.8     28.0     33.6     25.2   
Folate, DFE                       400.0    400.0    400.0    400.0    400.0   
Calcium, Ca                      1300.0   1300.0   1000.0   1000.0   1000.0   
Carbohydrate, by difference       130.0    130.0    130.0    130.0    130.0   
Iron, Fe                           15.0     11.0     18.0      8.0     18.0   
Magnesium, Mg                     360.0    410.0    310.0    400.0    320.0   
Niacin                             14.0     16.0     14.0     16.0     14.0   
Phosphorus, P                    1250.0   1250.0    700.0    700.0    700.0   
Potassium, K                     4700.0   4700.0   4700.0   4700.0   4700.0   
Riboflavin                          1.0      1.3      1.1      1.3      1.1   
Thiamin                             1.0      1.2      1.1      1.2      1.1   
Vitamin A, RAE                    700.0    900.0    700.0    900.0    700.0   
Vitamin B-12                        2.4      2.4      2.4      2.4      2.4   
Vitamin B-6                         1.2      1.3      1.3      1.3      1.3   
Vitamin C, total ascorbic acid     65.0     75.0     75.0     90.0     75.0   
Vitamin E (alpha-tocopherol)       15.0     15.0     15.0     15.0     15.0   
Vitamin K (phylloquinone)          75.0     75.0     90.0    120.0     90.0   
Zinc, Zn                            9.0     11.0      8.0     11.0      8.0   

                                M 31-50   F 51+   M 51+  
Nutrition                                                
Energy                           2200.0  1600.0  2000.0  
Protein                            56.0    46.0    56.0  
Fiber, total dietary               30.8    22.4    28.0  
Folate, DFE                       400.0   400.0   400.0  
Calcium, Ca                      1000.0  1200.0  1000.0  
Carbohydrate, by difference       130.0   130.0   130.0  
Iron, Fe                            8.0     8.0     8.0  
Magnesium, Mg                     420.0   320.0   420.0  
Niacin                             16.0    14.0    16.0  
Phosphorus, P              

In [49]:
# Maximum Requirement DataFrame
max_requirement = pd.read_csv("Dietary_Requirements_diet_maximums.csv")
max_requirement = max_requirement.drop(max_requirement.columns[1], axis = 1)
max_requirement = max_requirement.set_index("Nutrition")
max_requirement

C 1-3  F 4-8  M 4-8  F 9-13  M 9-13  F 14-18  M 14-18  F 19-30  \
Nutrition                                                                    
Sodium, Na   1500   1900   1900    2200    2200     2300     2300     2300   
Energy       2500   2500   2500    2800    3000     3100     3100     3100   

            M 19-30  F 31-50  M 31-50  F 51+  M 51+  
Nutrition                                            
Sodium, Na     2300     2300     2300   2300   2300  
Energy         3100     3100     3100   3100   3100

In [50]:
# Concatenate min_requirement and max_requirement to get Matrix b
matrix_b = pd.concat([min_requirement, -max_requirement])
matrix_b

C 1-3   F 4-8   M 4-8  F 9-13  M 9-13  \
Nutrition                                                                
Energy                          1000.0  1200.0  1400.0  1600.0  1800.0   
Protein                           13.0    19.0    19.0    34.0    34.0   
Fiber, total dietary              14.0    16.8    19.6    22.4    25.2   
Folate, DFE                      150.0   200.0   200.0   300.0   300.0   
Calcium, Ca                      700.0  1000.0  1000.0  1300.0  1300.0   
Carbohydrate, by difference      130.0   130.0   130.0   130.0   130.0   
Iron, Fe                           7.0    10.0    10.0     8.0     8.0   
Magnesium, Mg                     80.0   130.0   130.0   240.0   240.0   
Niacin                             6.0     8.0     8.0    12.0    12.0   
Phosphorus, P                    460.0   500.0   500.0  1250.0  1250.0   
Potassium, K                    3000.0  3800.0  3800.0  4500.0  4500.0   
Riboflavin                         0.5     0.6     0.6     0.9     0.9   
Thiamin                            0.5     0.6     0.6     0.9     0.9   
Vitamin A, RAE                   300.0   400.0   400.0   600.0   600.0   
Vitamin B-12                       0.9     1.2     1.2     1.8     1.8   
Vitamin B-6                        0.5     0.6     0.6     1.0     1.0   
Vitamin C, total ascorbic acid    15.0    25.0    25.0    45.0    45.0   
Vitamin E (alpha-tocopherol)       6.0     7.0     7.0    11.0    11.0   
Vitamin K (phylloquinone)         30.0    55.0    55.0    60.0    60.0   
Zinc, Zn                           3.0     5.0     5.0     8.0     8.0   
Sodium, Na                     -1500.0 -1900.0 -1900.0 -2200.0 -2200.0   
Energy                         -2500.0 -2500.0 -2500.0 -2800.0 -3000.0   

                                F 14-18  M 14-18  F 19-30  M 19-30  F 31-50  \
Nutrition                                                                     
Energy                           1800.0   2200.0   2000.0   2400.0   1800.0   
Protein                            46.0     52.0     46.0     56.0     46.0   
Fiber, total dietary               25.2     30.8     28.0     33.6     25.2   
Folate, DFE                       400.0    400.0    400.0    400.0    400.0   
Calcium, Ca                      1300.0   1300.0   1000.0   1000.0   1000.0   
Carbohydrate, by difference       130.0    130.0    130.0    130.0    130.0   
Iron, Fe                           15.0     11.0     18.0      8.0     18.0   
Magnesium, Mg                     360.0    410.0    310.0    400.0    320.0   
Niacin                             14.0     16.0     14.0     16.0     14.0   
Phosphorus, P                    1250.0   1250.0    700.0    700.0    700.0   
Potassium, K                     4700.0   4700.0   4700.0   4700.0   4700.0   
Riboflavin                          1.0      1.3      1.1      1.3      1.1   
Thiamin                             1.0      1.2      1.1      1.2      1.1   
Vitamin A, RAE                    700.0    900.0    700.0    900.0    700.0   
Vitamin B-12                        2.4      2.4      2.4      2.4      2.4   
Vitamin B-6                         1.2      1.3      1.3      1.3      1.3   
Vitamin C, total ascorbic acid     65.0     75.0     75.0     90.0     75.0   
Vitamin E (alpha-tocopherol)       15.0     15.0     15.0     15.0     15.0   
Vitamin K (phylloquinone)          75.0     75.0     90.0    120.0     90.0   
Zinc, Zn                            9.0     11.0      8.0     11.0      8.0   
Sodium, Na                      -2300.0  -2300.0  -2300.0  -2300.0  -2300.0   
Energy                          -3100.0  -3100.0  -3100.0  -3100.0  -3100.0   

                                M 31-50   F 51+   M 51+  
Nutrition                                                
Energy                           2200.0  1600.0  2000.0  
Protein                            56.0    46.0    56.0  
Fiber, total dietary               30.8    22.4    28.0  
Folate, DFE                       400.0   400.0   400.0  
Calcium, Ca

### [A] Objective Function ($p$)

In [51]:
# Normalized units with corresponding prices
p = food_price.drop(food_price.columns[0], axis = 1)
p = p[~p['FDC Price'].astype(str).str.contains('nan', case=False, na=False)]
# astype(str) converts 'FDC Price' column to strings and applies the str.contains method to filter out rows containing 'nan'. 

p = p.set_index('Name')
p_series = p["FDC Price"]
# str(p["FDC Price"].values[1])[:7]
list_p = []
for index in p_series:
    index = str(index)[:10]   # index is now string
    list_p.append(index)
    
p["FDC Price"] = list_p

p

FDC Price
Name                              
chicken breast tenders  1.76149347
Atlantic salmon fillet  2.42288026
Avocado                 1.41095847
Oat - Whole Grain       0.58705950
Honey                   1.76075860
Greek Yogurt Plain      0.77051560
Raw Almonds             1.76149347
Banana                  0.43871990
Spinach                 1.32056895
Quinoa                  0.83555197
Unsalted Butter         1.32056895
Parsley                 45.2514540
Ripe olives             3.11978596
Sweet Potato/Yam        0.43871990
Green Cabbage           0.32848877
Peanut Butter           0.86345465
Carrots                 0.32848877
Whole Wheat Bread       0.47940521
Long Grain rice         0.27447551
Spaghetti Pasta         0.55115565
Ground beef             1.10010668
Shredded cheese         1.10120899
Cheese Slices           1.75928885
Portabello Mushrooms    1.76369809
Tomatoes                1.30160919
Cucumber                0.66028447
Brussels Sprouts        0.87964442
Pork Loin Sirloin       0.43871990
Strawberries            1.32056895
Kidney(chicken)         0.88008535

### [A] Solution:

In [53]:
# This part outputs the least cost diet for different population group providing different restrictions in food list.

# If this code doen't work, run from the top again!!!!!!!!!!!!
gender = input("Gender(C or F or M): ").upper()
age_range = input("Age Range: ")  # Format of age_range: "M 31-50", "F 51+", "C 1-3"
query_char_1 = str(gender) + " " + str(age_range)

tolerance_level = 1e-6
result = lp(p, -matrix_A, -matrix_b[query_char_1], method='highs')
result
print("\n")
print("Cost of diet for population group of %s is $%4.2f per day." % (query_char_1,result.fun))

print("\n")

diet_recommandation = pd.Series(result.x,index = p.index)
print("\nThey'll be eating (in hectograms or milliliters):")
print(diet_recommandation[diet_recommandation >= tolerance_level])

Gender(C or F or M): f
Age Range: 19-30


Cost of diet for population group of F 19-30 is $6.34 per day.



They'll be eating (in hectograms or milliliters):
Name
Oat - Whole Grain    0.017531
Green Cabbage        0.576614
Peanut Butter        1.367148
Carrots              9.134905
Whole Wheat Bread    2.474360
Long Grain rice      1.274259
Shredded cheese      0.309206
Kidney(chicken)      0.097166
dtype: float64


### [B] Is Your Solution Edible?
> __The diet below is the least cost diet we picked. The main ingredients include:__
> * __Oat - Whole Grain: &emsp;  &emsp; &emsp;0.02 hectograms__
> * __Green Cabbage: &nbsp;&emsp; &emsp; &emsp;&emsp;0.58 hectograms__
> * __Peanut Butter:   &emsp;&emsp;  &emsp;&emsp;&emsp;1.37 hectograms__
> * __Carrots: &emsp; &emsp; &ensp; &emsp; &emsp;  &emsp;   &ensp;   &nbsp;     9.13 hectograms__
> * __Whole Wheat Bread:     &emsp;  &emsp; 2.47 hectograms__
> * __Long Grain rice:  &emsp; &ensp;  &emsp;  &emsp;    1.27 hectograms__
> * __Shredded cheese:     &nbsp; &ensp;  &emsp; &emsp;  0.31 hectograms__
> * __Kidney(chicken):  &emsp; &nbsp;  &emsp;  &emsp;  0.10 hectograms__


> __Additional Ingredients:__
> * __Sesame Oil: &emsp;  &emsp; &emsp;1 tablespoon__
> * __Soy Sause: &nbsp; &emsp; &emsp;&emsp;2 tablespoons__


__This meal costs $6.34 per day.__

__The pictures of our meal below  [created by AI:](https://copilot.microsoft.com/)__

<br>


<div style="display: flex; justify-content: center; align-items: center;">
  <img src="food_1.png" alt="First Image" style="width: 45%; margin-right: 10px;">
  <img src="food_2.png" alt="Second Image" style="width: 45%;">
</div>

<br>


__The code block right below shows how we got the outcome:__

In [58]:
# This part outputs the least cost diet for different population group providing different restrictions in food list.
# This part also outputs the comparision between the optimized outcome and the nutritional recommendation.

# If this code doen't work, run from the top again!!!!!!!!!!!!
tolerance_level = 1e-6
result = lp(p, -matrix_A, -matrix_b[query_char_1], method='highs')
result
print("\n")
print("The cost of %s is $%4.2f per day." % ("our diet",result.fun))

print("\n")

diet_recommandation = pd.Series(result.x,index = p.index)
print("\nWe'll be eating (in hectograms or milliliters):")
print(diet_recommandation[diet_recommandation >= tolerance_level])

print("\n")

tab = pd.DataFrame({"Outcome":np.abs(matrix_A).dot(diet_recommandation),"Recommendation":np.abs(matrix_b[query_char_1])})
print("\nWith the following nutritional outcomes of interest:")
tab



The cost of our diet is $6.34 per day.



We'll be eating (in hectograms or milliliters):
Name
Oat - Whole Grain    0.017531
Green Cabbage        0.576614
Peanut Butter        1.367148
Carrots              9.134905
Whole Wheat Bread    2.474360
Long Grain rice      1.274259
Shredded cheese      0.309206
Kidney(chicken)      0.097166
dtype: float64



With the following nutritional outcomes of interest:


Outcome  Recommendation
Nutrition                                                  
Energy                          3100.000000          2000.0
Protein                           92.137715            46.0
Fiber, total dietary              52.350525            28.0
Folate, DFE                      803.386405           400.0
Calcium, Ca                     1000.000000          1000.0
Carbohydrate, by difference      332.747672           130.0
Iron, Fe                          18.000000            18.0
Magnesium, Mg                    605.595883           310.0
Niacin                            49.344674            14.0
Phosphorus, P                   1579.242788           700.0
Potassium, K                    4700.000000          4700.0
Riboflavin                         1.315539             1.1
Thiamin                            2.526690             1.1
Vitamin A, RAE                  7630.528816           700.0
Vitamin B-12                       2.400000             2.4
Vitamin B-6                        2.636045             1.3
Vitamin C, total ascorbic acid    75.000000            75.0
Vitamin E (alpha-tocopherol)      15.000000            15.0
Vitamin K (phylloquinone)        183.830819            90.0
Zinc, Zn                          12.548481             8.0
Sodium, Na                      2300.000000          2300.0
Energy                          3100.000000          3100.0

### [B] Meal Reviews:
> __By providing the csv file of a specific food list below, the following code will return the estimated price, estimated nutrients and the comparisions with *Recommended Dietary Allowances (RDAs)*.__

In [59]:
# This part outputs the least cost diet for different population group providing different restrictions in food list.
# This part also outputs the comparision between the optimized outcome and the nutritional recommendation.

# If this code doen't work, run from the top again!!!!!!!!!!!!
Other_team_csv = str(input("Other team's diet csv file (e.g. Food_Price_Data.csv):"))
food_price_1 = pd.read_csv(Other_team_csv)  # It must be the same format as our "Food_Price_Data.csv"
food_price_1 = food_price_1.drop(food_price_1.columns[5:], axis = 1)
food_price_1["FDC Quantity"] = food_price_1.apply(lambda x: fdc.units(x["Quantity"], x["Unit"]), axis = 1)
food_price_1["FDC Price"] = food_price_1["Price ($)"] / food_price_1["FDC Quantity"]
warnings.filterwarnings("ignore", message="The unit of the quantity is stripped when downcasting to ndarray")
food_price_1 = food_price_1.drop(food_price_1.columns[2:6], axis = 1)  
print("\nThose are the ingredients in other team's diet:")
print(food_price_1)

print("\n")

matrix_A_1 = matrix_A_func(food_price_1)

tab_1 = pd.DataFrame({"Outcome":np.abs(matrix_A_1).dot(diet_recommandation),"Recommendation":np.abs(matrix_b["F 19-30"])})
print("\nWith the following nutritional outcomes of interest:")


result_1 = lp(p, -matrix_A_1, -matrix_b["F 19-30"], method='highs')

print("Cost of diet for the population group of %s is $%4.2f per day." % ("this team",result_1.fun))

print("\n")

tab_1

Other team's diet csv file (e.g. Food_Price_Data.csv):Food_Price_Data.csv

Those are the ingredients in other team's diet:
      fdcId                    Name                        FDC Price
0   2120006  chicken breast tenders   1.7614934748571718 / hectogram
1   2390897  Atlantic salmon fillet   2.4228802614118043 / hectogram
2   2094567    Cage free large eggs                 nan / milliliter
3   2344719                 Avocado   1.4109584779832163 / hectogram
4   2577037  Extra virgin olive oil                 nan / milliliter
5   2261421       Oat - Whole Grain   0.5870595095894454 / hectogram
6   2345841                   Honey    1.760758600649889 / hectogram
7   2067508      Greek Yogurt Plain    0.770515606336147 / hectogram
8   2608592             Raw Almonds   1.7614934748571718 / hectogram
9   2012128                  Banana   0.4387199017479063 / hectogram
10  2099349                 Spinach   1.3205689504874165 / hectogram
11  1858805                 Quinoa    0.835551973

Outcome  Recommendation
Nutrition                                                  
Energy                          3100.000000          2000.0
Protein                           92.137715            46.0
Fiber, total dietary              52.350525            28.0
Folate, DFE                      803.386405           400.0
Calcium, Ca                     1000.000000          1000.0
Carbohydrate, by difference      332.747672           130.0
Iron, Fe                          18.000000            18.0
Magnesium, Mg                    605.595883           310.0
Niacin                            49.344674            14.0
Phosphorus, P                   1579.242788           700.0
Potassium, K                    4700.000000          4700.0
Riboflavin                         1.315539             1.1
Thiamin                            2.526690             1.1
Vitamin A, RAE                  7630.528816           700.0
Vitamin B-12                       2.400000             2.4
Vitamin B-6                        2.636045             1.3
Vitamin C, total ascorbic acid    75.000000            75.0
Vitamin E (alpha-tocopherol)      15.000000            15.0
Vitamin K (phylloquinone)        183.830819            90.0
Zinc, Zn                          12.548481             8.0
Sodium, Na                      2300.000000          2300.0
Energy                          3100.000000          3100.0

### [C] Sensitivity of Solution:

In [56]:
# This part outputs the dynamic graph for the effects of price changes on subsistence diet cost.

# Effects of Price Changes on Subsistence Diet Cost
cf.go_offline()
scale = [.5,.6,.7,.8,.9,1.,1.1,1.2,1.3,1.4,1.5]
cost0 = result.fun
Price_response={}

for s in scale:
    cost = {}
    for i in range(len(p)):
        my_p = p.copy()
        my_p['FDC Price'] = pd.to_numeric(my_p['FDC Price'], errors='coerce')
        my_p.iloc[i] = my_p.iloc[i] * s 
        result_2 = lp(my_p, -matrix_A, -matrix_b[query_char_1], method='highs')
        cost[p.index[i]] = np.log(result_2.fun / cost0)
    Price_response[np.log(s)] = cost

Price_response = pd.DataFrame(Price_response).T

layout = {
    'title': 'Effects of Price Changes on Subsistence Diet Cost',
    'xaxis': {'title': 'Change in Log Price'},
    'yaxis': {'title': 'Change in Log Cost'}
}

Price_response.iplot(layout=layout)

### [B] What is Total Cost for Population of Interest?

In [57]:
# This part outputs the least cost diet for different population group providing different restrictions in food list.
# This part also outputs the comparision between the optimized outcome and the nutritional recommendation.

# We can also use this code block to analyze our project topic: Keto, Mediterranean and Vegan diets.
# If this code doen't work, run from the top again!!!!!!!!!!!!
Other_team_csv = str(input("Other team's diet csv file (e.g. Food_Price_Data.csv):"))
food_price_other_team = pd.read_csv(Other_team_csv)  # It must be the same format as our "Food_Price_Data.csv"
food_price_other_team = food_price_other_team.drop(food_price_other_team.columns[5:], axis = 1)
food_price_other_team["FDC Quantity"] = food_price_other_team.apply(lambda x: fdc.units(x["Quantity"], x["Unit"]), axis = 1)
food_price_other_team["FDC Price"] = food_price_other_team["Price ($)"] / food_price_other_team["FDC Quantity"]
warnings.filterwarnings("ignore", message="The unit of the quantity is stripped when downcasting to ndarray")
food_price_other_team = food_price_other_team.drop(food_price_other_team.columns[2:6], axis = 1)  
print("\nThose are the ingredients in other team's diet:")
print(food_price_other_team)

print("\n")

matrix_A_1 = matrix_A_func(food_price_other_team)

tab_1 = pd.DataFrame({"Outcome":np.abs(matrix_A_1).dot(diet_recommandation),"Recommendation":np.abs(matrix_b["F 19-30"])})
print("\nWith the following nutritional outcomes of interest:")


result_1 = lp(p, -matrix_A_1, -matrix_b["F 19-30"], method='highs')

print("Cost of diet for the population group of %s is $%4.2f per day." % ("this team",result_1.fun))

print("\n")

tab_1

Other team's diet csv file (e.g. Food_Price_Data.csv):Food_Price_Data.csv

Those are the ingredients in other team's diet:
      fdcId                    Name                        FDC Price
0   2120006  chicken breast tenders   1.7614934748571718 / hectogram
1   2390897  Atlantic salmon fillet   2.4228802614118043 / hectogram
2   2094567    Cage free large eggs                 nan / milliliter
3   2344719                 Avocado   1.4109584779832163 / hectogram
4   2577037  Extra virgin olive oil                 nan / milliliter
5   2261421       Oat - Whole Grain   0.5870595095894454 / hectogram
6   2345841                   Honey    1.760758600649889 / hectogram
7   2067508      Greek Yogurt Plain    0.770515606336147 / hectogram
8   2608592             Raw Almonds   1.7614934748571718 / hectogram
9   2012128                  Banana   0.4387199017479063 / hectogram
10  2099349                 Spinach   1.3205689504874165 / hectogram
11  1858805                 Quinoa    0.835551973

Outcome  Recommendation
Nutrition                                                  
Energy                          3100.000000          2000.0
Protein                           92.137715            46.0
Fiber, total dietary              52.350525            28.0
Folate, DFE                      803.386405           400.0
Calcium, Ca                     1000.000000          1000.0
Carbohydrate, by difference      332.747672           130.0
Iron, Fe                          18.000000            18.0
Magnesium, Mg                    605.595883           310.0
Niacin                            49.344674            14.0
Phosphorus, P                   1579.242788           700.0
Potassium, K                    4700.000000          4700.0
Riboflavin                         1.315539             1.1
Thiamin                            2.526690             1.1
Vitamin A, RAE                  7630.528816           700.0
Vitamin B-12                       2.400000             2.4
Vitamin B-6                        2.636045             1.3
Vitamin C, total ascorbic acid    75.000000            75.0
Vitamin E (alpha-tocopherol)      15.000000            15.0
Vitamin K (phylloquinone)        183.830819            90.0
Zinc, Zn                          12.548481             8.0
Sodium, Na                      2300.000000          2300.0
Energy                          3100.000000          3100.0

<br>

### __<div style="text-align: right"> <span style="font-family:Papyrus; ">-------------------------- The End.</span></div>__